### PR TITLE
Implements a transformation for axes tag propagation

### LINKED ISSUE #299: Introduce a pass for propagating axes tags
We could generalize the logic in https://github.com/kaushikcfd/meshmode/blob/62797c78447309c8ab86210e1fdc381a0bb64d05/meshmode/pytato_utils.py#L481 and make it available as a pytato transformation. I believe array axes corresponding to physical entities is a general enough characteristic of array programs to be shipped in `:mod:pytato`

### PR DESCRIPTION
**Pull request by kaushikcfd**
Closes #299.

Draft because:
- [x] Add tests
- [x] Pass CI
- [x] Docs
- [x] Includes commits from #224

### CODE REVIEW

File: pytato/transform/metadata.py [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+                lhs_var = self.get_var_for_axis(ary, iaxis)
+                rhs_var = self.get_var_for_tag(tag)
+                self.record_equation(lhs_var, rhs_var)
+
+    # }}}
+
+    # {{{ mapper interface
+
+    # type-ignore reason: signature not compatible with Mapper.rec
+    def rec(self, expr: ArrayOrNames) -> None:  # type: ignore
```
Comment (inducer) [inline pytato/transform/metadata.py:175]:
```suggestion
    def rec(self, expr: ArrayOrNames) -> None:  # type: ignore[override]
```
?

File: pytato/transform/metadata.py [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+    """
+    Returns a copy of *expr* with tags of type *tag_t* propagated along the
+    array operations with the tags propagation semantics implemented in
+    *equations_collector*.
+    """
+    equations_collector = equations_collector_t(tag_t)
+
+    equations_collector(expr)
+
+    # start BFS traversal with the literal as the sources.
```
Comment (inducer) [inline pytato/transform/metadata.py:664]:
```suggestion
    # start BFS traversal with the literals as the sources.
```

Comment (inducer) [inline pytato/transform/metadata.py:664]: Literal -> Tag? Known Tag? Above you've called the constants? Pick one terminology and stick to it?

File: pytato/transform/metadata.py [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+    # Defn. A Propagation graph is a graph where nodes denote variables and an
+    # edge between 2 nodes denotes an equality criterion.
+
+    propagation_graph = _get_propagation_graph_from_constaints(
+        equations_collector.equations)
+
+    constant_vars = frozenset(equations_collector.constant_to_var.values())
+    axis_to_solved_tags: Dict[Tuple[Array, int], Set[Tag]] = {}
+
+    for tag, var in equations_collector.constant_to_var.items():
```
Comment (inducer) [inline pytato/transform/metadata.py:675]: So this doesn't yell about inconsistencies, right? But I guess if something's a `UniqueTag`, it eventually will yell? At any rate, I think it's worth documenting this in more detail.

File: pytato/transform/metadata.py [hunk 4]
```diff
@@ -0,0 +1,687 @@
...
+        *,
+        tag_t: Type[Tag] = Tag,
+        equations_collector_t: Type[
+            AxesTagsEquationCollector] = AxesTagsEquationCollector,
+        unify_redn_descrs: bool = True,
+) -> ArrayOrNames:
+    """
+    Returns a copy of *expr* with tags of type *tag_t* propagated along the
+    array operations with the tags propagation semantics implemented in
+    *equations_collector*.
```
Comment (inducer) [inline pytato/transform/metadata.py:658]: Make this a clickable reference.

Comment (inducer) [inline pytato/transform/metadata.py:658]: The equations collector documents what edges will exist, but not what happens along an edge (and I'm not sure I get what is supposed to happen). See also below.

**Review by inducer — COMMENTED** [review]
Thanks for working on this! Some comments below.

File: pytato/transform/metadata.py (hunk 4 above)
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Not sure how I should do that. We typically refer to function arguments by italicizing them.

File: pytato/transform/metadata.py (hunk 3 above)
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: > But I guess if something's a UniqueTag, it eventually will yell? 

Yep, exactly!

> At any rate, I think it's worth documenting this in more detail.

Done!

File: pytato/transform/metadata.py (hunk 2 above)
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: "known tag", I think aptly, captures it. Used it.

Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

File: pytato/transform/metadata.py (hunk 1 above)
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

File: pytato/transform/metadata.py [hunk 5]
```diff
@@ -0,0 +1,687 @@
...
+        :class:`pytato.DistributedRecv` does not have any operands and so no
+        more equations are deduced.
+        """
+
+    # }}}
+
+# }}}
+
+
+def _get_propagation_graph_from_constaints(
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

File: pytato/transform/metadata.py [hunk 6]
```diff
@@ -0,0 +1,687 @@
...
+                self.record_equation(
+                    self.get_var_for_axis(expr.array, i_in_axis),
+                    self.get_var_for_axis(expr, i_out_axis))
+        # }}}
+
+    def map_reshape(self, expr: Reshape) -> None:
+        """
+        Reshaping generally does not preserve the axis between its input and
+        output and so no constraints are enforced except when the
+        :class:`pytato.Reshape` has come from a :func:`pytato.expand_dims`.
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py:457]: > Leading and trailing axes with matching length could conceivably be propagated.

That could also lead to unintended propagation between the inputs.

File: pytato/transform/metadata.py [hunk 7]
```diff
@@ -0,0 +1,687 @@
...
+        its corresponding axis of *expr*.
+        """
+        for bnd in expr.bindings.values():
+            self.rec(bnd)
+
+        try:
+            hlo = index_lambda_to_high_level_op(expr)
+        except UnknownIndexLambdaExpr:
+            # no propagation semantics implemented for such cases
+            return
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py:217]: Done!

File: pytato/transform/metadata.py [hunk 8]
```diff
@@ -0,0 +1,687 @@
...
+        propagation equations are recorded.
+        """
+
+    map_placeholder = _map_input_base
+    map_data_wrapper = _map_input_base
+    map_size_param = _map_input_base
+
+    def map_index_lambda(self, expr: IndexLambda) -> None:
+        """
+        The propagation semantics for a :class:`~pytato.IndexLambda` is
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

File: pytato/transform/metadata.py [hunk 9]
```diff
@@ -0,0 +1,687 @@
...
+        """
+        self.equations.append((lhs, rhs))
+
+    def record_equations_from_axes_tags(self, ary: Array) -> None:
+        """
+        Records equations for *ary*\'s axis tags of type :attr:`tag_t`.
+        """
+        for iaxis, axis in enumerate(ary.axes):
+            for tag in axis.tags_of_type(self.tag_t):
+                lhs_var = self.get_var_for_axis(ary, iaxis)
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

File: pytato/transform/metadata.py [hunk 10]
```diff
@@ -0,0 +1,687 @@
...
+            new_var = self.var_name_gen("")
+            self.axis_to_var[key] = new_var
+            return new_var
+
+    def get_var_for_tag(self, tag: Tag) -> str:
+        key = tag
+        try:
+            return self.constant_to_var[key]
+        except KeyError:
+            new_var = self.var_name_gen("c")
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done.

File: pytato/transform/metadata.py [hunk 11]
```diff
@@ -0,0 +1,687 @@
...
+
+    # {{{ unification helpers
+
+    def get_var_for_axis(self, ary: Array, iaxis: int) -> str:
+        key = (ary, iaxis)
+
+        try:
+            return self.axis_to_var[key]
+        except KeyError:
+            new_var = self.var_name_gen("")
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done.

File: pytato/transform/metadata.py [hunk 12]
```diff
@@ -0,0 +1,687 @@
...
+                    )
+                else:
+                    raise AssertionError
+
+    def map_concatenate(self, expr: Concatenate) -> None:
+        """
+        Records an equality equation between the axes of arrays being
+        concatenated and their corresponding axis in *expr*. No equation is
+        added for the concatenated axis i.e.
+        :attr:`pytato.array.Concatenate.axis`.
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py:301]: That's a fair question and I think that depends on the context of the tag. I don't think any default is perfect here.

File: pytato/transform/metadata.py [hunk 13]
```diff
@@ -0,0 +1,687 @@
...
+        """
+
+    map_placeholder = _map_input_base
+    map_data_wrapper = _map_input_base
+    map_size_param = _map_input_base
+
+    def map_index_lambda(self, expr: IndexLambda) -> None:
+        """
+        The propagation semantics for a :class:`~pytato.IndexLambda` is
+        implemented only for operations that can be risen to a
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py:202]: Done!

File: pytato/transform/metadata.py [hunk 14]
```diff
@@ -0,0 +1,687 @@
...
+    Users are encouraged to derive this mapper to implement domain-specific
+    axis tags propagation semantics.
+
+    .. attribute:: tag_t
+
+        The type of the tags that are to be propa
+
+    .. attribute:: equations
+
+        A :class:`list` of equation. Each equation is represented by 2-tuple
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done.

File: pytato/transform/metadata.py [hunk 15]
```diff
@@ -0,0 +1,687 @@
...
+        as ``("u", "v")`` that is mathematically interpreted as
+        :math:`\{u \doteq v\}`.
+
+    .. attribute:: constant_to_var
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Renamed to known_tag_to_var.

File: pytato/transform/metadata.py [hunk 16]
```diff
@@ -0,0 +1,687 @@
...
+
+        A mapping from an instance of :class:`Tag` to a :class:`str` it will be
+        referred to in :attr:`equations`.
+
+    .. attribute:: axis_to_var
+
+        A mapping from a :class:`tuple` of the form ``(array, iaxis)`` to the
+        :class:`str` it will be referred to in :attr:`equations`.
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

File: pytato/transform/metadata.py [hunk 17]
```diff
@@ -0,0 +1,687 @@
...
+    .. attribute:: equations
+
+        A :class:`list` of equation. Each equation is represented by 2-tuple
+        as ``("u", "v")`` that is mathematically interpreted as
+        :math:`\{u \doteq v\}`.
+
+    .. attribute:: constant_to_var
+
+        A mapping from an instance of :class:`Tag` to a :class:`str` it will be
+        referred to in :attr:`equations`.
```
[the remark this answers is not in the document]
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Done!

**Comment by kaushikcfd** [comment]
(There still some comments that I need to cater to, will request a review when I'm done addressing those)

File: pytato/transform/metadata.py (hunk 4 above)
Comment (kaushikcfd) [reply pytato/transform/metadata.py]: Added a bit more explanation here on how the propagation problem is reformulated as a unification problem in the axes tags.

**Comment by inducer** [comment]
LGTM, thanks!

**Pull request merged**